File: osuBeatmapTools.java
import java.lang.Math;
import java.io.*;

public class osuBeatmapTools {
    double Offset;
    int intOffset;
    String Area = "";
    String pt = "";
    String previousVolume = "100";
    String previousSample = "0";
    String previousType = "1";
    String previousInheret = "1";
    String beforeT = "";
    String currentT = "";
    int i = 1;
    int a, b;
    int countcheck = 0;

    public osuBeatmapTools() {
        try {
            // Step 1 : File Connection

            // CODE ARRAY:
            // General = 0
            // Editor = 1
            // Metadata = 2
            // Difficulty = 3
            // Events = 4
            // TimingPoints = 5
            // Colours = 6
            // HitObjects = 7
            File[] beatmap = new File[8];
            File input = new File("before.txt");
            beatmap[0] = new File("Sub/General.txt");
            beatmap[1] = new File("Sub/Editor.txt");
            beatmap[2] = new File("Sub/Metadata.txt");
            beatmap[3] = new File("Sub/Difficulty.txt");
            beatmap[4] = new File("Sub/Events.txt");
            beatmap[5] = new File("Sub/TimingPoints.txt");
            beatmap[6] = new File("Sub/Colours.txt");
            beatmap[7] = new File("Sub/HitObjects.txt");

            BufferedReader read = new BufferedReader(new FileReader(input));

            // STEP 1 : Separator
            String st;
            FileWriter[] write = new FileWriter[8];
            BufferedWriter[] write2 = new BufferedWriter[8];
            PrintWriter[] delete = new PrintWriter[8];
            for (int i = 0; i <= 7; i++) {
                write[i] = new FileWriter(beatmap[i], true);
                write2[i] = new BufferedWriter(write[i]);
                delete[i] = new PrintWriter(beatmap[i]);
                delete[i].close();
            }

            while ((st = read.readLine()) != null) {
                if (Area.equals("General")) {
                    write2[0].write(st);
                    write2[0].flush();
                    write2[0].newLine();
                }
                if (Area.equals("Editor")) {
                    write2[1].write(st);
                    write2[1].flush();
                    write2[1].newLine();
                }
                if (Area.equals("Metadata")) {
                    write2[2].write(st);
                    write2[2].flush();
                    write2[2].newLine();
                }
                if (Area.equals("Difficulty")) {
                    write2[3].write(st);
                    write2[3].flush();
                    write2[3].newLine();

                }
                if (Area.equals("Events")) {
                    write2[4].write(st);
                    write2[4].flush();
                    write2[4].newLine();
                }
                if (Area.equals("TimingPoints")) {
                    if (st.isEmpty() == false) {
                        String[] lineTimingPoint = st.split(",");
                        Offset = Double.parseDouble(lineTimingPoint[0]);
                        Offset = Math.floor(Offset);
                        intOffset = (int) Offset;
                        lineTimingPoint[0] = Integer.toString(intOffset);

                        if (lineTimingPoint[6].equals("1")) {
                            previousVolume = lineTimingPoint[5];
                            previousSample = lineTimingPoint[4];
                            previousType = lineTimingPoint[3];
                        }
                        if (lineTimingPoint[6].equals("0")) {
                            if (lineTimingPoint[3].equals("1") &&
                                    lineTimingPoint[4].equals("0") &&
                                    lineTimingPoint[5].equals("100")) {
                                lineTimingPoint[5] = previousVolume;
                                lineTimingPoint[4] = previousSample;
                                lineTimingPoint[3] = previousType;
                            }
                            previousVolume = lineTimingPoint[5];
                            previousSample = lineTimingPoint[4];
                            previousType = lineTimingPoint[3];
                        }

                        String temp = "";
                        for (int j = 0; j < 8; j++) {
                            temp = temp + lineTimingPoint[j];
                            if (j != 7) {
                                temp = temp + ",";
                            }
                        }

                        // REDUNDANT CHECK

                        if (beforeT.isEmpty() == false) {
                            String[] beforeTD = beforeT.split(",");
                            a = Integer.parseInt(beforeTD[0]);

                            String[] currentTD = temp.split(",");
                            b = Integer.parseInt(currentTD[0]);
                            if (beforeTD[1].equals(currentTD[1]) &&
                                    beforeTD[2].equals(currentTD[2]) &&
                                    beforeTD[3].equals(currentTD[3]) &&
                                    beforeTD[4].equals(currentTD[4]) &&
                                    beforeTD[5].equals(currentTD[5]) &&
                                    beforeTD[6].equals(currentTD[6]) &&
                                    beforeTD[7].equals(currentTD[7]) && a == b || a + 1 == b) {
                                st = temp;
                                beforeT = temp;
                                System.out.print(countcheck + " ");
                                countcheck++;

                            } else {

                                st = temp;
                                beforeT = temp;
                                write2[5].write(st);
                                write2[5].newLine();
                                write2[5].flush();
                            }
                        } else {

                            beforeT = st;
                            write2[5].write(st);
                            write2[5].newLine();
                            write2[5].flush();
                        }

                        i++;
                    }
                }
                if (Area.equals("Colours")) {
                    write2[6].write(st);
                    write2[6].flush();
                    write2[6].newLine();
                }
                if (Area.equals("HitObjects")) {
                    write2[7].write(st);
                    write2[7].flush();
                    write2[7].newLine();
                }

                if (st.equals("[General]")) {
                    Area = "General";
                    System.out.println(Area);
                }
                if (st.equals("[Editor]")) {
                    Area = "Editor";
                    System.out.println(Area);
                }
                if (st.equals("[Metadata]")) {
                    Area = "Metadata";
                    System.out.println(Area);
                }
                if (st.equals("[Difficulty]")) {
                    Area = "Difficulty";
                    System.out.println(Area);
                }
                if (st.equals("[Events]")) {
                    Area = "Events";
                    System.out.println(Area);
                }
                if (st.equals("[TimingPoints]")) {
                    Area = "TimingPoints";
                    System.out.println(Area);
                }
                if (st.equals("[Colours]")) {
                    Area = "Colours";
                    System.out.println(Area);
                }
                if (st.equals("[HitObjects]")) {
                    Area = "HitObjects";
                    System.out.println(Area);
                }
                if (st.isEmpty() == true) {
                    Area = "a";
                }
            }
        } catch (Exception e) {
            System.out.println("ERROR");
        }
        System.out.println(countcheck);
    }
}

File: osuBeatmapTools.py
import math

Redundant = 0

area = 'Default'
previousVolume = "100"
previousSample = "0"
previousType = "1"
previousInheret = "1"

beforeT = ""
currentT = ""
st = ""

File = ['','','','','','','','','']
File[0] = 'afterpy.osu'
File[1] = 'Sub/General.txt'
File[2] = 'Sub/Editor.txt'
File[3] = 'Sub/Metadata.txt'
File[4] = 'Sub/Difficulty.txt'
File[5] = 'Sub/Events.txt'
File[6] = 'Sub/TimingPoints.txt'
File[7] = 'Sub/Colours.txt'
File[8] = 'Sub/HitObjects.txt'

for x in range(9):
  open(File[x], 'w').close()

with open('before.osu', 'r', errors='ignore') as infile:
    for line in infile:
        line = line.strip()

        areas = {
        '[General]': 'General',
        '[Editor]': 'Editor',
        '[Metadata]': 'Metadata',
        '[Difficulty]': 'Difficulty',
        '[Events]': 'Events',
        '[TimingPoints]': 'TimingPoints',
        '[Colours]': 'Colours',
        '[HitObjects]': 'HitObjects'
        }

        if '[' in line:
            area = areas[line]
            # print(area)
        elif line == '':
            x = 0
        else:
            if area == 'Default':
                with open(File[0], 'a') as inFile:
                    inFile.write('osu file format v14 \n\n')
            elif area == 'General':
                with open(File[1], 'a') as inFile:
                    inFile.write(str(line)+'\n')
            elif area == 'Editor':
                with open(File[2], 'a') as inFile:
                    inFile.write(str(line)+'\n')
            elif area == 'Metadata':
                with open(File[3], 'a') as inFile:
                    inFile.write(str(line)+'\n')
            elif area == 'Difficulty':
                with open(File[4], 'a') as inFile:
                    inFile.write(str(line)+'\n')
            elif area == 'Events':
                with open(File[5], 'a') as inFile:
                    inFile.write(str(line)+'\n')
            elif area == 'TimingPoints':
                if line.strip():
                    lineTimingPoint = line.split(',')
                    rounded_number = math.floor(float(lineTimingPoint[0]))
                    lineTimingPoint[0] = str(rounded_number)

                    if lineTimingPoint[6] == '1':
                        previousVolume = lineTimingPoint[5]
                        previousSample = lineTimingPoint[4]
                        previousType = lineTimingPoint[3]
                        # newLineUnhiretion = ','.join(map(str, lineTimingPoint))
                        # with open(File[6], 'a') as inFile:
                        #     inFile.write(str(newLineUnhiretion)+'\n')
                        # lineTimingPoint[6] = 0
                        newLine = ','.join(map(str, lineTimingPoint))
                    elif lineTimingPoint[6] == '0':
                        if lineTimingPoint[3] == '1' and lineTimingPoint[4] == '0' and lineTimingPoint[5] == '100':
                            lineTimingPoint[5] = previousVolume
                            lineTimingPoint[4] = previousSample
                            lineTimingPoint[3] = previousType
                        previousVolume = lineTimingPoint[5]
                        previousSample = lineTimingPoint[4]
                        previousType = lineTimingPoint[3]
                        newLine = ','.join(map(str, lineTimingPoint))

                    # Redundant Check    
                    if beforeT == '':
                        with open(File[6], 'a') as inFile:
                            inFile.write(str(newLine)+'\n')
                        beforeT = newLine
                    elif beforeT != '':
                        beforeTD = beforeT.split(',')
                        currentTD = newLine.split(',')
                        if (beforeTD[1] == currentTD[1] and beforeTD[2] == currentTD[2] and beforeTD[3] == currentTD[3] and beforeTD[4] == currentTD[4] and beforeTD[5] == currentTD[5] and beforeTD[6] == currentTD[6] and beforeTD[7] == currentTD[7] ):
                            beforeT = newLine
                            Redundant = Redundant + 1
                        else:
                            beforeT = newLine
                            with open(File[6], 'a') as inFile:
                                inFile.write(str(newLine)+'\n')
            elif area == 'Colours':
                with open(File[7], 'a') as inFile:
                    inFile.write(str(line)+'\n')
            elif area == 'HitObjects':
                with open(File[8], 'a') as inFile:
                    inFile.write(str(line)+'\n')
print('Redundant = '+str(Redundant))

for y in range(1,9):
    strings = {
        1: '[General]\n',
        2: '[Editor]\n',
        3: '[Metadata]\n',
        4: '[Difficulty]\n',
        5: '[Events]\n',
        6: '[TimingPoints]\n',
        7: '[Colours]\n',
        8: '[HitObjects]\n',
    }

    with open(File[0], 'a') as inFile:
        inFile.write(strings[y])

    with open(File[y], 'r') as infile:
        for line in infile:
            line = line.strip()
            newLine = line

            with open(File[0], 'a') as inFile:
                inFile.write(str(newLine)+'\n')
    with open(File[0], 'a') as inFile:
                inFile.write('\n')
            